lxd/device/gpu_physical.go

```diff
@@ -98,10 +101,209 @@ func (d *gpuPhysical) Start() (*deviceConfig.RunConfig, error) {
 	return d.startContainer()
 }
 
+// startCDIDevices starts all the devices given in a CDI specification:
+// * `unix-char` (representing the card and non-card devices)
+// * `disk` (representing the mounts)).
+func (d *gpuPhysical) startCDIDevices(cdiID cdi.ID, configDevices cdi.ConfigDevices, runConf *deviceConfig.RunConfig) error {
+	srcFDHandlers := make([]*os.File, 0)
+	defer func() {
+		for _, f := range srcFDHandlers {
+			_ = f.Close()
+		}
+	}()
+
+	for _, conf := range configDevices.UnixCharDevs {
+		if conf["source"] == "" {
+			return fmt.Errorf("The source of the unix-char device %q used for CDI is empty", conf)
+		}
+
+		_, major, minor, err := unixDeviceAttributes(conf["source"])
+		if err != nil {
+			return err
+		}
+
+		// Here putting a 'cdi.unix' prefix with 'd.name' as a device name with create an directory entry like:
+		// <lxd_var_path>/devices/<instance_name>/cdi.unix.<gpu_device_name>.<path_encoded_relative_dest_path>
+		// 'unixDeviceSetupCharNum' is already checking for dupe entries so we have no validation to do here.
+		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "cdi.unix", d.name, conf, major, minor, conf["path"], false, runConf)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, conf := range configDevices.BindMounts {
+		if conf["source"] == "" {
+			return fmt.Errorf("The source of the disk device %q used for CDI is empty", conf)
+		}
+
+		srcPath := shared.HostPath(conf["source"])
+		destPath := conf["path"]
+		relativeDestPath := strings.TrimPrefix(destPath, "/")
+
+		// This time, the created path will be like:
+		// <lxd_var_path>/devices/<instance_name>/nvidia.com.<gpu_device_name>.<path_encoded_relative_dest_path>
```

Comment:
this needs updating i think?

Reply:
Something like this?
```go
// <lxd_var_path>/devices/<instance_name>/<cdi_vendor>.<gpu_device_name>.<path_encoded_relative_dest_path>
```

Reply:
In our MM discussions you showed an example:

```
cdi.unix.gpu0.dev-dri-card0
```

This example doesn't correlate with either the original comment or your proposed new one.

I notice also that in your example from MM, some are prefixed with "cdi." and others are prefixed with "nvidia.com." why is that?

Reply:
ah but this is different here: the bind mounts are prefixed with the CDI vendor name, not `cdi.unix`

Reply:
why not `cdi.disk`?

Reply:
As an aside, I notice that `removeUnixDevices` functions have hardcoded references to "infiniband.unix" which would benefit from using the constant `IBDevPrefix` - please can you update those?

Do we also need to update `removeUnixDevices` function with support for CDI unix devices?

Reply:
We should probably move `removeUnixDevices` into the common driver too rather than duplicating it for lxc and qemu instance drivers.

Reply:
yes why not. I can chage that.

Reply:
@tomponline  updated

Reply:
Thanks, sounds good!